crypto/evp/keymgmt_lib.c

```diff
@@ -0,0 +1,307 @@
+#include "internal/nelem.h"
+#include "internal/evp_int.h"
+#include "internal/asn1_int.h"
+#include "internal/provider.h"
+#include "evp_locl.h"
+
+static OSSL_PARAM *paramdefs_to_params(const OSSL_PARAM *paramdefs)
+{
+    size_t cnt;
+    const OSSL_PARAM *p;
+    OSSL_PARAM *params, *q;
+
+    for (cnt = 1, p = paramdefs; p->key != NULL; p++, cnt++)
+        continue;
+
+    params = OPENSSL_zalloc(cnt * sizeof(*params));
+
+    for (p = paramdefs, q = params; ; p++, q++) {
+        *q = *p;
+        if (p->key == NULL)
+            break;
+
+        q->data = NULL;          /* In case the provider used it */
+        q->return_size = 0;
+    }
+
+    return params;
+}
+
+typedef union align_block_un {
+    OSSL_UNION_ALIGN;
+} ALIGN_BLOCK;
+
+#define ALIGN_SIZE  sizeof(ALIGN_BLOCK)
+
+static void *allocate_params_space(OSSL_PARAM *params)
+{
+    unsigned char *data = NULL;
+    size_t space;
+    OSSL_PARAM *p;
+
+    for (space = 0, p = params; p->key != NULL; p++)
+        space += ((p->return_size + ALIGN_SIZE - 1) / ALIGN_SIZE) * ALIGN_SIZE;
+
+    data = OPENSSL_zalloc(space);
+
+    for (space = 0, p = params; p->key != NULL; p++) {
+        p->data = data + space;
+        space += ((p->return_size + ALIGN_SIZE - 1) / ALIGN_SIZE) * ALIGN_SIZE;
+    }
+
+    return data;
+}
+
+void *evp_keymgmt_export_to_provider(EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
+{
+    void *provkey = NULL;
+    size_t i, j;
+
+    /*
+     * If there is an underlying legacy key and it has changed, invalidate
+     * the cache of provider keys.
+     */
+    if (pk->pkey.ptr != NULL) {
+        /*
+         * If there is no dirty checker, this key can't be used
+         * with providers
+         */
+        if (pk->ameth->is_dirty == NULL)
+            return NULL;
+
+        if (pk->ameth->is_dirty(pk))
+            evp_keymgmt_clear_pkey_cache(pk);
+    }
+
+    /*
+     * See if we have exported to this provider already.
+     * If we have, return immediately.
+     */
+    for (i = 0;
+         i < OSSL_NELEM(pk->pkeys) && pk->pkeys[i].keymgmt != NULL;
+         i++) {
+        if (keymgmt == pk->pkeys[i].keymgmt)
+            return pk->pkeys[i].provkey;
+    }
+
+    if (pk->pkey.ptr != NULL) {
+        /* There is a legacy key, try to export that one to the provider */
+
+        /* If the legacy key doesn't have an export function, give up */
+        if (pk->ameth->export_to == NULL)
+            return NULL;
+
+        /* Otherwise, simply use it */
+        provkey = pk->ameth->export_to(pk, keymgmt);
+    } else {
+        /*
+         * Here, there is no legacy key, so we look at the already
+         * cached but there are cached provider keys, and import from
+         * the first that supports it (i.e. use its export function),
+         * and export the imported data to the new provider.
+         */
+
+        /*
+         * If the given keymgmt doesn't have an import function, give up
+         */
+        if (keymgmt->importkey_priv == NULL)
```

Comment:
So I'm trying to work out whether we actually *need* the import/export key_pub variants since we never seem to call them. Is it enough just to import/export a key, with the private fields being optional? Alternatively do we need to detect whether a key is private or public and then call the right function?

Reply:
Well, I guess it comes down to certainty...  when do you want to know that you have the right set of key material for your operation?  Typically, encrypt and verify requires a private key while decrypt and sign requires a public key.  Do you want to know what you have before engaging the operation, or is it enough that the operation comes back with a "Hey dude, this key doesn't look right"?

Reply:
Well, if we're after certainty then calling `importkey_priv` unconditionally in this code does not seem the right thing to do since it won't always be a private key. Presumably `importkey_priv` would be expected to fail if we only passed it a public key. I'm tempted to say that its ok that its not until you attempt the operation that you hit the problem. After all that's basically what we have now - which would suggest we don't need different variants for priv vs pub.

Reply:
Ok, that's a good point ('cause you're right, when a `EVP_PKEY_assign_RSA` is done, we have no idea what's in the assigned legacy key).

Will fix

Reply:
This still seems to be an issue?

Reply:
Thanks for the reminder.

Reply:
Done